### ISSUE TITLE
No mails are sent when providers offer AUTH PLAIN authentification only

### ISSUE THREAD

**Issue by wnhre2ur8cxx8** [passbolt/passbolt_api#350]
## Unable to send Emails to some providers

* Passbolt Version: e848cd4d9ef8982e405ee8350fca325e2c562fad
* Platform and Target:
-- Operating system: Debian Buster (10)
-- PHP: 7.3.14-1~deb10u1
-- Web server: nginx/1.14.2
-- Database server: mariadb  Ver 15.1
-- etc.: Vserver

### What you did
I installed passbolt with the script, everything was working as expected and went very smooth. Except the email sending part. I startet a forum thread with my problems here: https://community.passbolt.com/t/emails-not-sending/2771 

First I thought it was the same with every provider, then I found a provider that worked for me (without TLS and instead with the ssl:// URL) Still the provider I want to use doesnt not work in passbolt. I tried connecting to it manually and in the end did so successfully. From the same system and the same user/pass combination. See [this post](https://community.passbolt.com/t/emails-not-sending/2771/5) for an exact output of a successful connection.

### What happened

Although my email is working in thunderbird/webmail/manual smtp it is not working in passbolt.
```
# sudo -u www-data /var/www/passbolt/bin/cake passbolt  send_test_email --recipient=my@email.com

     ____                  __          ____  
    / __ \____  _____ ____/ /_  ____  / / /_ 
   / /_/ / __ `/ ___/ ___/ __ \/ __ \/ / __/ 
  / ____/ /_/ (__  |__  ) /_/ / /_/ / / /    
 /_/    \__,_/____/____/_.___/\____/_/\__/   

 Open source password manager for teams
---------------------------------------------------------------
 Debug email shell
---------------------------------------------------------------

Email configuration
---------------------------------------------------------------
Host: mail.systemli.org
Port: 587
Username: myworkingusername
Password: *********
TLS: true

Sending email from: myemailname
Sending email to: my@email.com
---------------------------------------------------------------

Trace
[220] mail1.systemli.org ESMTP Postfix (Debian/GNU)
> EHLO localhost
[250] mail1.systemli.org
[250] PIPELINING
[250] SIZE 40960000
[250] ETRN
[250] STARTTLS
[250] ENHANCEDSTATUSCODES
[250] 8BITMIME
[250] DSN
[250] CHUNKING
> STARTTLS
[220] 2.0.0 Ready to start TLS
> EHLO localhost
[250] mail1.systemli.org
[250] PIPELINING
[250] SIZE 40960000
[250] ETRN
[250] AUTH PLAIN
[250] AUTH=PLAIN
[250] ENHANCEDSTATUSCODES
[250] 8BITMIME
[250] DSN
[250] CHUNKING

A test email could not be sent.
Error: SMTP Error: 535 5.7.8 Error: authentication failed: Invalid authentication mechanism
```

### What you expected to happen
I expected a successful mail transfer.

**Comment by stripthis** [comment]
Can you share your postfix configuration? Maybe someone with more experience can find some hint there. Could it be because the way the credentials are sent does not match the kind that is accepted from your postfix configuration?

**Comment by wnhre2ur8cxx8** [comment]
Unfortunately I cannot send you the configuration, because I am not hosting it. The way I was able to connect is documented in the mentioned thread. But to recap, I entered:

- `openssl s_client -starttls smtp -connect mail.systemli.org:587`
- `EHLO there`
- `AUTH PLAIN`
- in a second terminal I entered: `echo -ne '\0my@working.mail\0mypassword'| base64
- I copy pasted the base64 result into the SMTP session
- Profit: `235 2.7.0 Authentication successful`

**Comment by stripthis** [comment]
This is not my area of expertise but I suspect the mailer system use something else to format the credentials like `DIGEST-MD5 CRAM-MD5`, rather than plain base64. Is there a way of contacting your provider to know what kind of configuration they have / accept?

**Comment by wnhre2ur8cxx8** [comment]
Yes I can try. But doesnt my manual tests show that base64 is indeed working? I am also not an expert  as you can probably tell :D

**Comment by stripthis** [comment]
My hypothesis is maybe that `PLAIN` base64 encoding is not supported by the library which we use, maybe because there is a more popular more secure alternative. It's just an hypothesis.

**Comment by wnhre2ur8cxx8** [comment]
Indeed I am digging through the code and it seems that in `vendor/cakephp/cakephp/src/Mailer/Transport/SmtpTransport.php` the _auth function is using `AUTH LOGIN ` instead of ÀUTH PLAIN` which leads to the exact same error message when I try it manually. `535 5.7.8 Error: authentication failed: Invalid authentication mechanism`

I will investigate what the difference is :) thanks for the hint

**Comment by wnhre2ur8cxx8** [comment]
This is simply not implemented in the Mailer library. It should be super simple though. I created a ticket with the relevant information: https://github.com/cakephp/cakephp/issues/14634

**Comment by wnhre2ur8cxx8** [comment]
FYI: cakephp will probably add PLAIN support in 3.8-3.9. So it would be nice to update passbolts requirements for this feature.

I also did a small hack to get it running for now: https://github.com/wnhre2ur8cxx8/cakephp/blob/smtp_auth_plain/src/Mailer/Transport/SmtpTransport.php Put the file in vendor/cakephp/cakephp/src/Mailer/Transport if you need some hacky PLAIN support :D (and dont forget to backup the original file)

**Comment by garrettboone** [comment]
@wnhre2ur8cxx8 base64 is encoding, not encryption, just to be clear. What are you hoping to achieve by using base64? If authentication using an encryption standard, then base64 is not that category. It's great for changing the format of data in order to transmit via web requests, but for authentication, you would want to use an encryption standard TLSv1.2 or higher.

When you set TLS to false, that is PLAIN auth, without base64. Changing your username into something that is not your username will make it impossible to authenticate into your account, assuming that your provider allows PLAIN auth.

**Comment by wnhre2ur8cxx8** [comment]
> @wnhre2ur8cxx8 base64 is encoding, not encryption, just to be clear. What are you hoping to achieve by using base64? If authentication using an encryption standard, then base64 is not that category. It's great for changing the format of data in order to transmit via web requests, but for authentication, you would want to use an encryption standard TLSv1.2 or higher.

I dont understand what you are referring to, unfortunately. I never had any issue with base64 encoding. Maybe you are confusing "PLAIN AUTH" with somethig else. It is is a specific mechanism to authenticate yourself via SMTP. It has nothing to do with TLS or anything.

> When you set TLS to false, that is PLAIN auth, without base64. Changing your username into something that is not your username will make it impossible to authenticate into your account, assuming that your provider allows PLAIN auth.

You are mistaken. The CakePHP library in the version passbolt is using is not able to authenticate via PLAIN AUTH. Again it has nothing to do with TLS. PLAIN support was very recently added to cakephp: https://github.com/cakephp/cakephp/releases/tag/4.1.0-RC1 also https://github.com/cakephp/cakephp/issues/14634

 It would be nice if you could update the requirements for passbolt when the next stable release gets ready :)

**Comment by wnhre2ur8cxx8** [comment]
Regarding PLAIN AUTH: https://www.samlogic.net/articles/smtp-commands-reference-auth.htm

**Comment by garrettboone** [comment]
I'm saying base64 has nothing to do with email authentication. I was curious why you are using it and what you were hoping it would do for you. Your code seems to focus on it, but why?

**Comment by stripthis** [comment]
> It would be nice if you could update the requirements for passbolt when the next stable release gets ready :)

It's actually more than updating the composer requirements. Cakephp v4 is a major upgrade and we need to convert our plugins, the third party plugin we use, change the minimal php requirements, etc. I just wanted to flag this as we have other priorities at the moment (custom fields, folders, react migration, UI improvements, etc.)

**Comment by wnhre2ur8cxx8** [comment]
They (cakephp devs) wrote about integrating this into 3.9, the issue was also part of the 3.9 Milestone. So maybe it will be included there as well. I will tell you here :stuck_out_tongue:

**Comment by stripthis** [comment]
That's nice of them! So, we will ship it when they backport it.

**Comment by wnhre2ur8cxx8** [comment]
https://github.com/cakephp/cakephp/releases/tag/3.8.13

PLAIN support is included in the latest 3.8 release.

**Comment by stripthis** [comment]
We'll include the update in the next release.

**Comment by stripthis** [comment]
Fixed in v2.13.

**Comment by wnhre2ur8cxx8** [comment]
@stripthis This is not fixed in 2.13. I think you forgot to add cakephp 3.8.13 to the composer.lock file or something? Anyways the library is still at 3.18.12 and has no working AUTH PLAIN support.

**Comment by cedricalfonsi** [comment]
Hello @wnhre2ur8cxx8 
Thanks for the report, it will be fixed with the next release.

**Comment by stripthis** [comment]
Fixed with v3.0.0
